src/ray/core_worker/core_worker.cc

```diff
@@ -459,15 +453,43 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       {{"worker_id", worker_id.Hex()}});
 }
 
+CoreWorker::~CoreWorker() { RAY_LOG(INFO) << "Core worker is destructed"; }
+
 void CoreWorker::Shutdown() {
-  io_service_.stop();
+  RAY_LOG(INFO) << "Shutting down a core worker.";
+  is_shutdown_ = true;
   if (options_.worker_type == WorkerType::WORKER) {
+    // Running in a main thread.
     direct_task_receiver_->Stop();
     task_execution_service_.stop();
   }
   if (options_.on_worker_shutdown) {
+    // Running in a main thread.
     options_.on_worker_shutdown(GetWorkerID());
   }
+
+  // Running in a io_thread_.
+  io_service_.stop();
+  RAY_LOG(INFO) << "Waiting for joining a core worker io thread.";
+  if (io_thread_.joinable()) {
+    io_thread_.join();
+  }
+  // Stop gcs client first since it runs in io_thread_.
+  if (gcs_client_) {
```

Comment:
This is the main fix. We should call `Disconnect`, which resets some pointer after io thread is drained (joined).